File: world/world.go
package world

import (
	"github.com/harrisonthorne/chicky-chicky-go/blocks"
	"github.com/harrisonthorne/chicky-chicky-go/maths"
	"github.com/harrisonthorne/chicky-chicky-go/render"

	"math/rand"
)

const MaxWorldSize = 256 // in chunks. 256 yields thousands and thousands of blocks in each direction

// World contains a slice of Plots
type World struct {
	terrain [][][]blocks.Chunk
	seed    i64

	heightGradientVectors      [MaxWorldSize][MaxWorldSize]maths.Vec2
	humidityGradientVectors    [MaxWorldSize][MaxWorldSize]maths.Vec2
	temperatureGradientVectors [MaxWorldSize][MaxWorldSize]maths.Vec2

	caveGradientVectors [MaxWorldSize][MaxWorldSize][MaxWorldSize]maths.Vec3
	oreGradientVectors  [MaxWorldSize][MaxWorldSize][MaxWorldSize]maths.Vec3

	renderDistance i32
}

fn NewWorld(seed i64) *World {
	w = &World{
		seed: seed,
	}

	w.heightGradientVectors = generate2DGradientMap(seed)
	w.humidityGradientVectors = generate2DGradientMap(seed + 1)
	w.temperatureGradientVectors = generate2DGradientMap(seed + 2)

	w.caveGradientVectors = generate3DGradientMap(seed + 3)
	w.oreGradientVectors = generate3DGradientMap(seed + 4)

	return w
}

fn generate2DGradientMap(seed i64) [MaxWorldSize][MaxWorldSize]maths.Vec2 {
	r = rand.New(rand.NewSource(i64(seed)))
	m = [MaxWorldSize][MaxWorldSize]maths.Vec2{}

	for x = 0; x < len(m); x++ {
		for y = 0; y < len(m[x]); y++ {
			m[x][y] = maths.Vec2{r.f32(), r.f32()}
		}
	}

	return m
}

fn generate3DGradientMap(seed i64) [MaxWorldSize][MaxWorldSize][MaxWorldSize]maths.Vec3 {
	r = rand.New(rand.NewSource(i64(seed)))
	m = [MaxWorldSize][MaxWorldSize][MaxWorldSize]maths.Vec3{}

	for x = 0; x < len(m); x++ {
		for y = 0; y < len(m[x]); y++ {
			for z = 0; z < len(m[x][y]); z++ {
				m[x][y][z] = maths.Vec3{r.f32(), r.f32(), r.f32()}
			}
		}
	}

	
	return m
}

// Stringify serializes the world into a big string that can be parsed to
// recreate the world.
fn (w *World) Stringify() string {
	return ""
}

fn (w *World) generateChunk(x, y, z i32) {

}

fn (w *World) render(c *render.Camera) {
	eyeChunkX = i32(c.Position().X / (blocks.BlockWidth * blocks.ChunkSize))
	eyeChunkY = i32(c.Position().Y / (blocks.BlockWidth * blocks.ChunkSize))
	eyeChunkZ = i32(c.Position().Z / (blocks.BlockWidth * blocks.ChunkSize))

	for x = eyeChunkX - w.renderDistance; x < eyeChunkX+w.renderDistance; x++ {
		if x < 0 || x >= len(w.terrain) {
			continue
		}
		for y = eyeChunkY - w.renderDistance; y < eyeChunkY+w.renderDistance; y++ {
			if y < 0 || y >= len(w.terrain[x]) {
				continue
			}
			for z = eyeChunkZ - w.renderDistance; z < eyeChunkZ+w.renderDistance; z++ {
				if z < 0 || z >= len(w.terrain[x][y]) {
					continue
				}
				w.terrain[x][y][z].render(c)
			}
		}
	}
}

File: src/world/world.rs
use blocks;
use maths;
use render;
use rand;

const MAX_WORLD_SIZE: i32 = 1024; // in chunks. 1024 yields thousands upon thousands of blocks in each direction

// World contains a slice of Plots
struct World {
    terrain: VecDeque<VecDeque<VecDeque<Chunk>>>,
    seed:    i64,

    height_gradient_vectors:      [[Vec2; MAX_WORLD_SIZE]; MAX_WORLD_SIZE],
    humidity_gradient_vectors:    [[Vec2; MAX_WORLD_SIZE]; MAX_WORLD_SIZE],
    temperature_gradient_vectors: [[Vec2; MAX_WORLD_SIZE]; MAX_WORLD_SIZE],

    caveGradientVectors: [[[Vec3; MAX_WORLD_SIZE]; MAX_WORLD_SIZE]; MAX_WORLD_SIZE],
    oreGradientVectors:  [[[Vec3; MAX_WORLD_SIZE]; MAX_WORLD_SIZE]; MAX_WORLD_SIZE],
    render_distance: i32,
}

impl World {
    fn new(seed: i64) -> Self {
        let height_gradient_vectors = generate_2d_gradient_map(seed);
        let humidity_gradient_vectors = generate_2d_gradient_map(seed + 1);
        let temperature_gradient_vectors = generate_2d_gradient_map(seed + 2);

        let cave_gradient_vectors = generate_3d_gradient_map(seed + 3);
        let ore_gradient_vectors = generate_3d_gradient_map(seed + 4);

        Self {
            terrain: Vec::new(),
            seed,
            height_gradient_vectors,
            humidity_gradient_vectors,
            temperature_gradient_vectors,
            cave_gradient_vectors,
            ore_gradient_vectors,
        }
    }

    /// Serializes the world into a big String that can be parsed to
    /// recreate the world.
    pub fn stringify(&self) -> String {
        return ""
    }

    pub fn generate_chunk(&self, x, y, z i32) {

    }

    fn render(&self, c *render.Camera) {
        eyeChunkX = c.Position().X / (blocks.BLOCK_WIDTH * blocks.CHUNK_SIZE) as i32;
        eyeChunkY = c.Position().Y / (blocks.BLOCK_WIDTH * blocks.CHUNK_SIZE) as i32;
        eyeChunkZ = c.Position().Z / (blocks.BLOCK_WIDTH * blocks.CHUNK_SIZE) as i32;

        for x in (eyeChunkX - w.renderDistance)..(eyeChunkX+w.renderDistance) {
            if x < 0 || x >= w.terrain.len() {
                continue;
            }
            for y in (eyeChunkY - w.renderDistance)..(eyeChunkY+w.renderDistance) {
                if y < 0 || y >= w.terrain[x].len() {
                    continue;
                }
                for z in (eyeChunkZ - w.renderDistance)..(eyeChunkZ+w.renderDistance) {
                    if z < 0 || z >= w.terrain[x][y].len() {
                        continue;
                    }
                    w.terrain[x][y][z].render(c);
                }
            }
        }
    }
}


fn generate_2d_gradient_map(seed: i64) -> [[Vec2; MAX_WORLD_SIZE]; MAX_WORLD_SIZE] {
    r = rand.new(rand.new_source(seed)) as i64;
    m = [MAX_WORLD_SIZE][MAX_WORLD_SIZE]maths.Vec2{};

    for x in 0..m.len(); x++ {
        for y in 0..m[x].len(); y++ {
            m[x][y] = maths.Vec2{r.f32(), r.f32()}
        }
    }

    return m
}

fn generate_3d_gradient_map(seed: i64) -> [MAX_WORLD_SIZE][MAX_WORLD_SIZE][MAX_WORLD_SIZE]maths.Vec3 {
    let r = rand.new(rand.newSource(seed)) as i64;
    let m = [MAX_WORLD_SIZE][MAX_WORLD_SIZE][MAX_WORLD_SIZE]maths.Vec3{};

    for x in 0..m.len() {
        for y in 0..m[x].len() {
            for z 0..m[x][y].len() {
                m[x][y][z] = maths.Vec3{r.f32(), r.f32(), r.f32()}
            }
        }
    }
}
